mysys/mf_iocache.c

```diff
@@ -106,21 +121,78 @@ init_functions(IO_CACHE* info)
       programs that link against mysys but know nothing about THD, such
       as myisamchk
     */
+    DBUG_ASSERT(!(info->myflags & MY_ENCRYPT));
     break;
   case SEQ_READ_APPEND:
     info->read_function = _my_b_seq_read;
-    info->write_function = 0;			/* Force a core if used */
+    DBUG_ASSERT(!(info->myflags & MY_ENCRYPT));
+    break;
+  case READ_CACHE:
+    if (info->myflags & MY_ENCRYPT)
+    {
+      DBUG_ASSERT(info->share == 0);
+      info->read_function= _my_b_encr_read;
+      break;
+    }
+    /* fall through */
+  case WRITE_CACHE:
+    if (info->myflags & MY_ENCRYPT)
+    {
+      info->write_function= _my_b_encr_write;
+      break;
+    }
+    /* fall through */
```

Comment:
Why fall through here? Either read/write_function will be initialised already, either the DBUG_ASSERT below will fire

Reply:
Reworked by separating "encrypted" / "non-encrypted" code paths to make it more readable.